File: 213.php
<?php

class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function rob($nums) {
        if (empty($nums)) {
            return 0;
        }

        $length = count($nums);
        if ($length == 1) {
            return $nums[0];
        }

        $dp1[0] = $nums[0];
        $dp1[1] = max($nums[1], $nums[0]);

        for ($i = 2; $i < $length - 1; $i++) {
            $dp1[$i] = max($nums[$i] + $dp1[$i - 2], $dp1[$i - 1]);
        }

        $dp2[1] = $nums[1];
        $dp2[2] = max($nums[1], $nums[2]);
        for ($i = 3; $i < $length; $i++) {
            $dp2[$i] = max($nums[$i] + $dp2[$i - 2], $dp2[$i - 1]);
        }

        return max($dp1[$length - 2], $dp2[$length - 1]);
    }
}

$solution = new Solution();
var_dump($solution->rob([1, 3, 1, 3, 100]));

File: go/dp/213.go
package main

import "fmt"

func rob(nums []int) int {
	length := len(nums)
	if length == 1 {
		return nums[0]
	}
	if length == 2 {
		return max(nums[0], nums[1])
	}

	dp0 := getDp(nums[0 : len(nums)-1])
	dp1 := getDp(nums[1:])

	return max(dp0, dp1)
}

func getDp(nums []int) int {
	dp0, dp1 := nums[0], max(nums[0], nums[1])
	for i := 2; i < len(nums); i++ {
		v := max(dp0+nums[i], dp1)
		dp0, dp1 = dp1, v
	}
	return dp1
}

func main() {
	nums := []int{1, 2, 3, 1}
	fmt.Println(rob(nums))
}
